build-info-extractor-gradle/src/main/groovy/org/jfrog/gradle/plugin/artifactory/extractor/listener/ArtifactoryDependencyResolutionListener.groovy

```diff
@@ -0,0 +1,92 @@
+package org.jfrog.gradle.plugin.artifactory.extractor.listener
+
+import org.apache.commons.lang.ArrayUtils
+import org.gradle.api.artifacts.DependencyResolutionListener
+import org.gradle.api.artifacts.ModuleVersionIdentifier
+import org.gradle.api.artifacts.ResolvableDependencies
+import org.gradle.api.artifacts.result.DependencyResult
+import org.gradle.api.artifacts.result.ResolvedComponentResult
+import org.gradle.api.artifacts.result.ResolvedDependencyResult
+
+import static org.jfrog.build.extractor.BuildInfoExtractorUtils.getModuleIdString
+
+/**
+ * Represents a DependencyResolutionListener, used to populate a dependency hierarchy map for each dependency in each module,
+ * which is used in the 'requestedBy' field of every dependency in the build info.
+ * Does so by listening to the 'afterResolve' event of every module.
+ */
+class ArtifactoryDependencyResolutionListener implements DependencyResolutionListener {
+    final Map<String, Map<String, String[][]>> modulesHierarchyMap = new HashMap()
+
+    @Override
+    void beforeResolve(ResolvableDependencies dependencies) {
+    }
+
+    @Override
+    void afterResolve(ResolvableDependencies dependencies) {
+        if (dependencies.getResolutionResult().getAllDependencies().isEmpty()) {
+            return
+        }
+        updateModulesHierarchyMap(dependencies)
+    }
+
+    /**
+     * Handles the modules' hierarchy map update.
+     * @param dependencies - Module's resolved dependencies.
+     */
+    void updateModulesHierarchyMap(ResolvableDependencies dependencies) {
+        String compId = getGav(dependencies.getResolutionResult().getRoot().getModuleVersion())
+        // If the module was already visited, update it.
+        Map<String, String[][]> hierarchyMap = (Map<String, String[][]>) modulesHierarchyMap.get(compId)
+        if (hierarchyMap == null) {
+            hierarchyMap = new HashMap<>()
+        }
+        updateDependencyMap(hierarchyMap, dependencies.getResolutionResult().getAllDependencies())
+        modulesHierarchyMap.put(compId, hierarchyMap)
+    }
+
+    /**
+     * Iterates over each resolved dependency and updates the map with it's parents.
+     * @param hierarchyMap - Dependency map of the module.
+     * @param dependencies - Set of the resolved dependencies.
+     */
+    private void updateDependencyMap(Map<String, String[][]> hierarchyMap, Set<DependencyResult> dependencies) {
+        for (DependencyResult dependency : dependencies) {
+            // Update the map for every resolved dependency.
+            if (dependency instanceof ResolvedDependencyResult) {
+                String[] newDependent = getPathToRoot(dependency)
+
+                // Add the new dependent's path to root to the 2d array.
+                String compId = getGav(dependency.getSelected().getModuleVersion())
+                String[][] curDependants = hierarchyMap[compId]
+                curDependants = ArrayUtils.add(curDependants, newDependent)
```

Comment:
This will only happen once for every dependency. The dependencies are not expected to have duplicates, so for every dependency we will add a dependant array once in this loop. The `ArrayUtils.add()` is used to add the new dependant array to ones found in previous triggers of this listener.